### PR TITLE
KAFKA-14471: Move IndexEntry and related to storage module

### PR DESCRIPTION
**Pull request by ijuma**
For broader context on this change, please check:

* KAFKA-14470: Move log layer to storage module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by ijuma** [comment]
@dengziming do you happen to have cycles to review this PR?

**Comment by dengziming** [comment]
> @dengziming do you happen to have cycles to review this PR?

Sure, I have added it to my review backlog.

File: storage/src/main/java/org/apache/kafka/server/log/internals/IndexEntry.java [hunk 1]
```diff
@@ -14,7 +14,9 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+package org.apache.kafka.server.log.internals;
 
-package kafka.log
-
-class CorruptIndexException(message: String) extends RuntimeException(message)
```
Comment (dengziming) [inline storage/src/main/java/org/apache/kafka/server/log/internals/IndexEntry.java:20]: It's weird that git thinks CorruptIndexException is rename to IndexEntry.

File: storage/src/main/java/org/apache/kafka/server/log/internals/OffsetPosition.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            return false;
+
+        OffsetPosition that = (OffsetPosition) o;
+
+        return offset == that.offset
+            && position == that.position;
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (dengziming) [inline storage/src/main/java/org/apache/kafka/server/log/internals/OffsetPosition.java:57]: Scala case class will generated same methods, right?

**Review by dengziming — APPROVED** [review]
LGTM, the 2 comments are only trivial questions.

File: storage/src/main/java/org/apache/kafka/server/log/internals/IndexEntry.java (hunk 1 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/IndexEntry.java:20]: Yes, I agree. Not sure we can do anything about this though.

File: storage/src/main/java/org/apache/kafka/server/log/internals/OffsetPosition.java (hunk 2 above)
Comment (ijuma) [reply storage/src/main/java/org/apache/kafka/server/log/internals/OffsetPosition.java:57]: Correct, case classes automatically generate `toString`, `hashCode` and `equals` and hence why I did the same here. However, the underlying implementation is not the same, but that's not an issue for us.

**Comment by ijuma** [comment]
JDK 8 build passed while there was 1 unrelated failure for the each of the other JDKs:

> Build / JDK 11 and Scala 2.13 / testBumpTransactionalEpoch(String).quorum=kraft – kafka.api.TransactionsTest
> 1m 18s
> Build / JDK 17 and Scala 2.13 / testTransactionAfterTransactionIdExpiresButProducerIdRemains(String).quorum=zk – kafka.api.ProducerIdExpirationTest

Merging to trunk.

**Pull request merged**